lib/src/stream_subscription_transformer.dart

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'async_memoizer.dart';
+import 'delegate/stream_subscription.dart';
+
+typedef Future _CancelHandler<T>(StreamSubscription<T> inner);
+
+typedef void _SimpleHandler<T>(StreamSubscription<T> inner);
+
+/// Creates a [StreamTransformer] that modifies the behavior of subscriptions to
+/// a stream.
+///
+/// When [StreamSubscription.cancel], [StreamSubscription.pause], or
+/// [StreamSubscription.resume] is called, the corresponding handler is invoked.
+/// By default, handlers just forward to the underlying subscription.
+///
+/// This guarantees that once the transformed [StreamSubscription] has been
+/// canceled, no further callbacks will be invoked. The [handlePause] and
```

Comment:
> The "This" here sounds like it refers to the entire previous paragraph, which doesn't guarantee that, e.g., if your cancel handler doesn't forward the cancel. If it only refers to the default, then it's not really interesting.
> 
> Should it say something like: 
> > It is guaranteed that ... 
> 
> ?
> (Probably not, since the code doesn't guarantee that).

Changed to just "Guarantees", which more clearly indicates that the function call itself provides the guarantee.

> So, it should be documented *very* heavily that user provided handlers MUST ensure that the inner stream doesn't provide any events while the outer subscription is paused or after it is cancelled.
> If they fail to do that, anything can happen. Failing the Stream contract can make code using `async for` completely unpredictable.

Done.

> Perhaps add a test after `handleCancel`/`handlePause` calls that `_inner` is paused or cancelled. (You can test `isPaused`, but to detect the cancel, you might need to wrap inner too, so that might not be worth it).
> Otherwise you can make sure that events happening while paused/cancelled cause async errors instead of being delivered.

I'd rather not add an additional layer of wrapping necessary to verify this.